### PR TITLE
Add annotation_processing_tool (kapt | javac) option to Kotlin modules

### PR DESCRIPTION
**Pull request by artem-zinnatullin**
Original motivation was explained in [discussion in Slack](https://buckbuild.slack.com/archives/C6N81MF55/p1533090203000161):

>Hey guys, after spending couple days on trying to switch us to Buck upstream I came to conclusion that `kapt` support needs to be changed in the upstream, `kapt` should be _optional_.
>
>We have mixed Java/Kotlin modules where we deliberately run annotation processing only against Java sources (our reasoning is build speed & stability in both Buck and Gradle setups).
>
>In our setup `kapt` is **not enabled** on the Gradle side
>
>On the Buck side however `kapt` kicks in even if we don't ask for it, what's worse it becomes primary method of annotation processing for mixed modules.
>
>We have 2 problems because of that:
>
>1) If module doesn't have any `.kt` with annotations that need apt, build fails because `.zip` with generated classes is not produced. Can be "solved" by converting one of the classes with annotations to Kotlin, but we don't need nor want that.
>
>2) Build will behave differently if we convert (see point №1) one of the classes to `.kt`, it'll work in Buck but won't work in Gradle because we don't use `kapt` there
>
>I've dealt with it before by turning `kapt` off in our fork, but we really want to get on to the upstream Buck to move faster, and `kapt` support is last huge blocker I'm dealing with…
>
>I think OkBuck could look for `kapt` configuration and only then enable `kapt` in Buck

@cwoodwar6 mentioned they have same problem:

>To your first point, we face the same issue and have been putting simple stub kotlin files to work around it. But we don’t like this solution either and have looking to fix it

---

Techincal aspects of this PR:

PR adds a separate `annotation_processing_tool = "kapt/javac"` to the DSL of Kotlin modules.

Default value is kept as `kapt` for compatibility.

Tools:

- `kapt` tool: Allows to run Java annotation processors against Kotlin sources while backporting it for Java sources too.
- `javac` tool: Works only against Java sources, Kotlin sources won't have access to generated classes at compile time.

---

If this gets accepted, the next step will be to update OkBuck with respecting Gradle `kapt` vs `annotationProcessor` configurations.

cc @thalescm @cwoodwar6 @kageiit

### CODE REVIEW

File: src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java [hunk 1]
```diff
@@ -261,20 +262,33 @@ public void createCompileStep(
...
+    switch (annotationProcessingTool) {
+      case KAPT:
+        finalJavacOptions = javacOptions.withAnnotationProcessingParams(AnnotationProcessingParams.EMPTY);
+        break;
+
+      case JAVAC:
+        finalJavacOptions = javacOptions;
+        break;
+
+        default:
```
Comment (cwoodwar6) [inline src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java]: nit: extra indent

File: src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java [hunk 2]
```diff
@@ -162,8 +162,23 @@ public BuildRule createBuildRule(
...
+    /**
+     * Works only against Java sources, Kotlin sources won't have access to generated
+     * classes at compile time.
+     */
+    JAVAC,
+  }
+
   public interface CoreArg extends JavaLibraryDescription.CoreArg {
     ImmutableList<String> getExtraKotlincArguments();
+    Optional<AnnotationProcessingTool> getAnnotationProcessingTool();
```
Comment (cwoodwar6) [inline src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java]: Be sure to document this in the rules. E.g. `docs/rule/android_library.soy`

File: src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java (hunk 1 above)
Comment (artem-zinnatullin) [reply src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java]: fixed

File: src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java (hunk 2 above)
Comment (artem-zinnatullin) [reply src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java]: added 👍

File: test/com/facebook/buck/jvm/kotlin/testdata/kotlin_library_description/com/example/ap/annotation-processing-tool-javac/AJavaClass.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.example.ap;
+
+class AJavaClass {
+  public void foo() {
+    new JavaClassWithJavaAnnotation_();
+    //new KotlinClassWithJavaAnnotation_();
```
Comment (thalescm) [inline test/com/facebook/buck/jvm/kotlin/testdata/kotlin_library_description/com/example/ap/annotation-processing-tool-javac/AJavaClass.java]: ✂️ ?

File: test/com/facebook/buck/jvm/kotlin/testdata/kotlin_library_description/com/example/ap/annotation-processing-tool-javac/AKotlinClass.kt [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.example.ap
+
+class AKotlinClass {
+    fun foo() {
+        //JavaClassWithJavaAnnotation_()
```
Comment (thalescm) [inline test/com/facebook/buck/jvm/kotlin/testdata/kotlin_library_description/com/example/ap/annotation-processing-tool-javac/AKotlinClass.kt]: ✂️ ?

File: test/com/facebook/buck/jvm/kotlin/testdata/kotlin_library_description/com/example/ap/annotation-processing-tool-javac/AJavaClass.java (hunk 3 above)
Comment (artem-zinnatullin) [reply test/com/facebook/buck/jvm/kotlin/testdata/kotlin_library_description/com/example/ap/annotation-processing-tool-javac/AJavaClass.java]: 👍

**Comment by styurin** [comment]
Could you reformat files and make sure CI passes?

**Comment by artem-zinnatullin** [comment]
Should be good now

File: docs/rule/android_library.soy [hunk 5]
```diff
@@ -159,6 +159,18 @@ compiled class files and resources.
   {/param}
 {/call}
 
+{call buck.arg}
+  {param name: 'annotation_processing_tool' /}
+  {param default : 'kapt' /}
+  {param desc}
+  Specifies the tool to use for annotation processing. Possible values: "kapt" or "javac".
+  "kapt" allows to run Java annotation processors against Kotlin sources while backporting
```
Comment (philipjameson) [inline docs/rule/android_library.soy]: s/"kapt" allows to run Java annotation/"kapt" allows running Java annotation/

File: docs/rule/kotlin_library.soy [hunk 6]
```diff
@@ -60,6 +60,18 @@ of the compiled class files and resources.
   {/param}
 {/call}
 
+{call buck.arg}
+  {param name: 'annotation_processing_tool' /}
+  {param default : 'kapt' /}
+  {param desc}
+  Specifies the tool to use for annotation processing. Possible values: "kapt" or "javac".
+  "kapt" allows to run Java annotation processors against Kotlin sources while backporting
```
Comment (philipjameson) [inline docs/rule/kotlin_library.soy]: s/"kapt" allows to run Java annotation/"kapt" allows to running Java annotation/

**Comment by artem-zinnatullin** [comment]
yeah, I was running tests before I opened up the PR

I remember I did some renames and refactorings in tests right before opening up the PR, must have had copied that `kotlin_library` twice into the BUCK file and didn't run tests after that, my bad

**Comment by artem-zinnatullin** [comment]
Thanks for review, everyone 😽

**Pull request closed without merging**